### PR TITLE
fix(just): widen to the unscoped suite when a diff outgrows argv

### PR DESCRIPTION
**Pull request by kixelated**
## Summary

- **Root cause.** `_changed` prints the changed-file list, and every hop of the dispatch takes it as *one* argument: `just _tools "$files"`, `just js check "$files"`, `just rs check-changed "$files"` -> `just rs _select "$FILES"`. just exports recipe parameters into the child's environment, so the list travels as a single execve string, and Linux caps one string at `MAX_ARG_STRLEN` (32 pages = 131072 bytes) however large `ARG_MAX` is (macOS caps the total at 1 MiB). Past that, every hop dies with `E2BIG`, which just reports as exit code 126 and "Argument list too long" -- naming neither the diff nor the recipe that could not receive it. PR #3225 hit this locally and in both CI jobs after accidentally including a 24k-file fuzz corpus, which buried the real problem.
- **Fix.** `_changed-cap` reads the list on stdin and prints the `ALL` sentinel the repo already uses when it does not fit in one argument. `check`, `fix`, and `test default` branch on it into `check-all` / `fix-all` / `test all`, the paths that pass no list at all.
- **Why not batching.** Each consumer needs the *whole* list to make one scoping decision (the `grep -qE` language-scope tests, `_select`'s seed -> dependent walk), so a chunked list would silently under-select. Widening never under-checks, and a diff that large selects most of the workspace anyway.
- **Why stdin.** It is both the only channel that can carry an oversized list and the only honest way to measure one: `${#var}` counts characters under a UTF-8 locale while execve counts bytes, so a path set of 3-byte characters would read as a third of its real size.
- **Why 64 KiB.** Half of `MAX_ARG_STRLEN`, leaving headroom for the rest of argv and env. Roughly 1500 paths. The fast path for normal diffs is unchanged.

## Public API changes

None. Build tooling and docs only; no Rust or TS surface is touched.

## Test plan

New `_changed-test`, wired into `_check-common` so it runs in both `check` and `check-all`, following the existing `_select-test` idiom. It drives `_changed-cap` with synthetic stdin fixtures, so it behaves identically in a dirty worktree, a clean checkout, and CI:

- over budget must print `ALL`; at budget and empty must not;
- the smallest nonempty list (1 byte) must exceed the only budget below it (0);
- one 3-byte CJK character must exceed a 2-byte budget -- the budget counts bytes, not characters;
- a non-numeric budget must be rejected, *and* that rejection must reach the caller rather than yielding an unbudgeted list;
- the budget must sit under Linux's `MAX_ARG_STRLEN`;
- a payload of exactly the budget must survive a real `just` argv hop (the operation that was failing);
- end to end, an over-budget real diff must print `ALL` -- guarded on there being a diff at all, and budgeted at 0.

Verified it bites: `just _changed-test 200000` fails on the `MAX_ARG_STRLEN` assertion.

Reproduced the original failure by creating 30k untracked files (1.8 MB of paths). Before: `just check`, `just test`, and `just fix` all exited 126 with "Argument list too long". After: all three print `changed: 1800032 bytes of paths exceeds the 65536 budget; selecting everything` and proceed into the unscoped suite.

Also verified the fast path is untouched -- a normal diff prints the same list as before, an empty diff still yields empty output and the "nothing changed" message, and the pre-existing "root orchestration changed" fallback still fires.

## Review follow-ups

Four findings, all in the *test* rather than the fix, each reproduced before being fixed and each given regression coverage:

1. **`_changed-test` depended on the checkout having a diff** (Codex, then CodeRabbit). `cache.yml` runs `just check-all` on a clean `main` checkout to warm the repo's only shared Rust cache, where the list is empty; the assertion would have failed that job. Fixed in 5931590 by splitting `_changed-cap` out and driving it with synthetic input.
2. **The budget counted characters, not bytes** (CodeRabbit). `${#files}` returned 3 for a 9-byte string, so UTF-8 paths under-counted 3x and could sail past a budget they actually blow. Fixed in c9708979 by measuring on stdin with `wc -c`.
3. **A rejected budget was swallowed.** Inside `[[ ]]` a command substitution's exit status is discarded, so `_changed` returned an unbudgeted list and exited 0 -- straight back into the E2BIG this prevents. Fixed in 31dfdeb.
4. **The end-to-end assertion budgeted at 1** (Codex). A checkout whose only change is a one-character root path is a 1-byte list, and `1 > 1` is false. Fixed in 8067b45 by budgeting at 0, which is below every nonempty list.

## Cross-Package Sync

No rows apply (no wire format, `moq-ffi`, or CLI surface changed). `CLAUDE.md` gains one line documenting the fallback, since it is the file that describes the diff-aware scoping.

(Written by Claude Opus 5)

### CODE REVIEW

File: justfile [hunk 1]
```diff
@@ -79,10 +86,71 @@ _changed $BASE:
...
+[private]
+_changed-test $LIMIT=changed_max:
+    #!/usr/bin/env bash
+    set -euo pipefail
+
+    fail() { echo "changed: _changed-test: $1" >&2; exit 1; }
+
+    # Any real diff is oversized against a one-byte budget, so this exercises the
+    # fallback without laying down a synthetic 30k-file corpus to trigger it.
+    [[ "$(just _changed "" 1)" == ALL ]] || fail "a diff over budget must print ALL"
```
[the remark this answers is not in the document]
Comment (kixelated) [reply justfile]: Good catch, and confirmed: `cache.yml:104` runs `just check-all` on a clean `main` checkout, where the list is empty, so `_changed` printed nothing and the assertion aborted `_check-common` before the suite ran. Fixed in 5931590 by splitting the size decision into `_changed-cap`, a pure function of a byte count, so the test drives it with synthetic sizes (`100 99`, `99 99`, `0 99`) instead of the ambient diff; the end-to-end assertion stays but is now guarded on there actually being a diff. Reproduced the exact condition (clean tree, base == HEAD): the old assertion fails, the new test passes, and the 30k-file fallback still works.

(Written by Claude Opus 5)

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: justfile [hunk 2]
```diff
@@ -79,10 +86,71 @@ _changed $BASE:
...
+    # fit in a single execve string. Linux caps one string at MAX_ARG_STRLEN (32
+    # pages, 131072 bytes) however large ARG_MAX is, so past that each hop dies
+    # with E2BIG, which just reports as exit code 126 and no mention of the diff.
+    # A diff that large selects most of the workspace anyway, so say `ALL` and
+    # let the caller widen to the unscoped suite, which passes no list at all.
+    [[ "$LIMIT" =~ ^[0-9]+$ ]] || {
+    	echo "changed: LIMIT must be a byte count, got: $LIMIT" >&2
+    	exit 2
+    }
+    if ((${#files} > LIMIT)); then
```
[the remark this answers is not in the document]
Comment (kixelated) [reply justfile]: Confirmed and fixed in c9708979e. `${#files}` counted 3 for a 9-byte string on this host, so a path set of 3-byte characters read as a third of its real size and could sail past a budget it actually blows. `_changed-cap` now takes the list on stdin and measures it with `wc -c` -- stdin being both the only channel that can carry an oversized list and the only honest way to measure one. Added a regression test (one CJK character, 3 bytes, against a 2-byte budget) that fails under the old character count.

(Written by Claude Opus 5)

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: justfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (kixelated) [reply justfile]: Agreed, and already fixed in 5931590 -- this review ran against edd1e, before that commit. `_changed-test` now drives `_changed-cap` with a synthetic list on stdin instead of the checkout's diff, so both the empty-diff case (cache.yml runs `just check-all` on a clean `main` checkout) and the exact-boundary case are covered deterministically. The end-to-end assertion is kept but guarded on there actually being a diff. Verified against the real clean-checkout condition: tree clean, base == HEAD, list empty, test passes.

(Written by Claude Opus 5)

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: justfile [hunk 3]
```diff
@@ -79,10 +86,120 @@ _changed $BASE:
...
+
+    # The budget still has to survive the hop it was sized for, which the checks
+    # above cannot show: they never pass a list that big to anything.
+    payload=$(head -c "$LIMIT" /dev/zero | tr '\0' x)
+    [[ "$(just _echo "$payload" | wc -c)" -eq $((LIMIT + 1)) ]] \
+    	|| fail "a $LIMIT-byte list does not survive an argv hop"
+
+    # End to end, but only when there is a diff to be oversized: see above.
+    if [[ -n "$(just _changed "" 100000000)" ]]; then
+    	[[ "$(just _changed "" 1)" == ALL ]] || fail "a diff over budget must print ALL"
```
[the remark this answers is not in the document]
Comment (kixelated) [reply justfile]: Confirmed and fixed in 8067b45f0. Reproduced your exact case (base == HEAD, sole untracked path `x`): the list is exactly 1 byte, `1 > 1` is false, so `_changed` returned `x` and the assertion failed. The end-to-end check now budgets at 0, which is below every nonempty list, and I added a synthetic assertion (`printf 'x' | _changed-cap 0` must print ALL) so the boundary is covered without depending on what the checkout holds. Verified the old budget fails and the new one passes in that same checkout.

(Written by Claude Opus 5)

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by kixelated** [comment]
@codex review

Head is now 8067b45f0. Since your review of 31dfdeba1 the only change is the one you asked for: the end-to-end cap assertion budgets at 0 instead of 1 (a one-character root path is a one-byte list that a one-byte budget does not exceed), plus a synthetic assertion so that boundary no longer depends on the checkout. Please confirm the delta before this lands.

(Written by Claude Opus 5)
